doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1748,31 +1757,39 @@ Controlling the proof flow
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction.  If
+   :n:`@one_term_with_bindings` is not provided, the tactic first does a
+   :tacn:`intros`
+   then succeeds if there is a hypothesis :n:`False` in the updated context, or
+   if there is a pair of hypotheses where one is simple negation of the other
+   (i.e. :n:`P` and not :n:`~P`).
 
-   .. exn:: No such assumption.
-      :undocumented:
+   Use the :tacn:`discriminate` tactic to prove the current goal when there
+   is a hypothesis with an impossible structural equality such as
+   :n:`0 = 1`.
 
-.. tacv:: contradiction @ident
+   If :n:`@one_term_with_bindings` has the type :n:`False`, the tactic
+   proves the current goal.  If the type is a negation, e.g. :n:`~P`, then
+   the current goal is replaced with :n:`P`.  (However, the converse doesn't
+   work, i.e. if the type is :n:`P` and the goal is :n:`~P`.)
 
-   The proof of False is searched in the hypothesis named :n:`@ident`.
+   .. todo PR: examples?  Or maybe it's simple enough that they're not needed
+
+   .. exn:: No such assumption.
+      :undocumented:
 
 .. tacn:: contradict @ident
-   :name: contradict
 
    This tactic allows manipulating negated hypothesis and goals. The name
    :n:`@ident` should correspond to a hypothesis. With :n:`contradict H`, the
    current goal and context is transformed in the following way:
 
-   + H:¬A ⊢ B becomes ⊢ A
-   + H:¬A ⊢ ¬B becomes H: B ⊢ A
+   .. todo PR: what is this below with ¬ and ⊢?  Wrong characters?
```

Comment:
These are just Unicode representations of negation (the non-Unicode variant in Coq is `~`, but `¬` is more universal in math / logic) and of a context + goal (the left-hand side is the context and the right-hand side is the goal, exactly as if it was a horizontal line in Coq). Readers with any logical background (even minimal) should be able to understand this.

Reply:
IIUC the standard meaning of A ⊢ B is "B is provable from A".  In Coq, the meaning is more "We hope to prove B from A, but we don't yet know if that's possible."

I'd rather not introduce new notation that only seems to used in this one spot, but it would take much more space to write these on multiple lines.  `cic.rst` uses turnstile a lot, I think with the second meaning.

Reply:
Indeed. Note that `spéciales` is allows you instantiating parameters not in order if you use `with`.

Reply:
> IIUC the standard meaning of A ⊢ B is "B is provable from A". In Coq, the meaning is more "We hope to prove B from A, but we don't yet know if that's possible."

These are not different things, when considered from the perspective of a logician / proof theorist.  `A ⊢ B` is an assertion that can itself be true or false (or proven or refuted).

Reply:
I think turnstile is used with varying meanings.  See [here](https://en.wikipedia.org/wiki/Turnstile_(symbol)#:~:text=In%20Unicode%2C%20the%20turnstile%20symbol,assertion%20sign%20(%E2%8A%A6).)).  In several cases it states something that is true rather than being a proposition that can be true or false.  But we don't have to agree on this since I say how to interpret it here (i.e. syntactically).